File: algorithms/medium/reordered_power_of_2.rs
use std::collections::HashMap;

fn count(mut n: i32) -> HashMap<i32, i32> {
    let mut ct = HashMap::new();
    while n > 0 {
        *ct.entry(n % 10).or_insert(0) += 1;
        n /= 10;
    }
    ct
}

impl Solution {
    pub fn reordered_power_of2(n: i32) -> bool {
        let m = count(n);
        (0..=32).any(|i| count(2_i32.pow(i)) == m)
    }
}

File: algorithms/medium/reordered_power_of_2.ex
defmodule Solution do
  @spec reordered_power_of2(n :: integer) :: boolean
  def reordered_power_of2(n) do
    m = count(n)

    0..32
    |> Enum.any?(fn i ->
      count(Integer.pow(2, i)) == m
    end)
  end

  def count(n), do: count(n, %{})
  def count(0, ct), do: ct

  def count(n, ct) do
    count(div(n, 10), Map.update(ct, rem(n, 10), 1, &(&1 + 1)))
  end
end
